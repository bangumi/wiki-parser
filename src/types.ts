export interface Wiki {
  type: string;
  data: WikiItem[];
}

/** JS 的 map 会按照插入顺序排序 */
export interface WikiMap {
  type: string;
  data: Map<string, WikiItem>;
}

export type WikiItemType = 'array' | 'object';

export class WikiArrayItem {
  k?: string;
  v?: string;

  constructor(k?: string, v?: string) {
    this.k = k || undefined;
    this.v = v;
  }
}

export class WikiItem {
  key: string;
  value?: string;
  array?: boolean;
  values?: WikiArrayItem[];

  constructor(key: string, value: string, type: WikiItemType) {
    this.key = key;
    switch (type) {
      case 'array': {
        this.array = true;
        this.values = [];
        break;
      }
      case 'object': {
        this.value = value;
        break;
      }
    }
  }

  private convertToArray() {
    if (this.array) {
      return;
    }

    this.array = true;
    this.values ??= [];

    this.values.push(new WikiArrayItem(undefined, this.value));
    this.value = undefined;
  }

  push(item: WikiItem) {
    this.convertToArray();

    if (item.array) {
      this.values?.push(...(item.values ?? []));
    } else {
      this.values?.push(new WikiArrayItem(undefined, item.value));
    }
  }
}
