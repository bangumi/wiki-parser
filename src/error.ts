export const GlobalPrefixError = "missing prefix '{{Infobox' at the start";
export const GlobalSuffixError = "missing suffix '}}' at the end";
export const ArrayNoCloseError = "array should be closed by '}'";
export const ArrayItemWrappedError = "array item should be wrapped by '[]'";
export const ExpectingNewFieldError = "missing '|' to start a new field";
export const ExpectingSignEqualError = "missing '=' to separate field name and value";

export class WikiSyntaxError extends Error {
  lino: number;
  line: string | null;

  constructor(lino: number, line: string | null, message: string) {
    super(toErrorString(lino, line, message));
    this.line = line;
    this.lino = lino;
  }
}

function toErrorString(lino: number, line: string | null, msg: string): string {
  if (line === null) {
    return `WikiSyntaxError: ${msg}, line ${lino}`;
  }

  return `WikiSyntaxError: ${msg}, line ${lino}: ${JSON.stringify(line)}`;
}
