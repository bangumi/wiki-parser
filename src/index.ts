import {
  ArrayItemWrappedError,
  ArrayNoCloseError,
  ExpectingNewFieldError,
  ExpectingSignEqualError,
  GlobalPrefixError,
  GlobalSuffixError,
  WikiSyntaxError,
} from './error';
import { prefix, suffix } from './shared';
import type { Wiki, WikiItemType, WikiMap } from './types';
import { WikiArrayItem, WikiItem } from './types';

export * from './types';
export * from './error';
export { stringify } from './stringify';

/** 解析 wiki 文本，以 `Map` 类型返回解析结果。 会合并重复出现的 key */
export function parseToMap(s: string): WikiMap {
  const w = parse(s);

  const data = new Map<string, WikiItem>();

  for (let item of w.data) {
    const previous = data.get(item.key);
    if (!previous) {
      data.set(item.key, item);
      continue;
    }

    previous.push(item);
  }

  return { type: w.type, data };
}

export function parse(s: string): Wiki {
  const wiki: Wiki = {
    type: '',
    data: [],
  };

  const strTrim = s.trim().replace(/\r\n/g, '\n');

  if (strTrim === '') {
    return wiki;
  }

  if (!strTrim.startsWith(prefix)) {
    throw new WikiSyntaxError(null, null, GlobalPrefixError);
  }
  if (!strTrim.endsWith(suffix)) {
    throw new WikiSyntaxError(null, null, GlobalSuffixError);
  }

  const arr = strTrim.split('\n');
  if (arr[0]) {
    wiki.type = parseType(arr[0]);
  }

  /* split content between {{Infobox xxx and }} */
  const fields = arr.slice(1, -1);

  let inArray = false;
  for (let i = 0; i < fields.length; ++i) {
    const line = fields[i]?.trim();
    const lino = i + 2;

    if (!line) {
      continue;
    }
    /* new field */
    if (line.startsWith('|')) {
      if (inArray) {
        throw new WikiSyntaxError(lino, line, ArrayNoCloseError);
      }
      const meta = parseNewField(lino, line);
      inArray = meta[2] === 'array';
      const field = new WikiItem(...meta);
      wiki.data.push(field);
      /* is Array item */
    } else if (inArray) {
      if (line.startsWith('}')) {
        inArray = false;
        continue;
      }
      if (i === fields.length - 1) {
        throw new WikiSyntaxError(lino, line, ArrayNoCloseError);
      }
      wiki.data[wiki.data.length - 1]?.values?.push(
        new WikiArrayItem(...parseArrayItem(lino, line)),
      );
    } else {
      throw new WikiSyntaxError(lino, line, ExpectingNewFieldError);
    }
  }
  return wiki;
}

const parseType = (line: string): string => {
  if (!line.includes('}}')) {
    return line.slice(prefix.length).trim();
  }
  return line.slice(prefix.length, line.indexOf('}}')).trim();
};

const parseNewField = (lino: number, line: string): [string, string, WikiItemType] => {
  const str = line.slice(1);
  const index = str.indexOf('=');

  if (index === -1) {
    throw new WikiSyntaxError(lino, line, ExpectingSignEqualError);
  }

  const key = str.slice(0, index).trim();
  const value = str.slice(index + 1).trim();
  switch (value) {
    case '{': {
      return [key, '', 'array'];
    }
    default: {
      return [key, value, 'object'];
    }
  }
};

const parseArrayItem = (lino: number, line: string): [string, string] => {
  if (!line.startsWith('[') || !line.endsWith(']')) {
    throw new WikiSyntaxError(lino, line, ArrayItemWrappedError);
  }
  const content = line.slice(1, -1);
  const index = content.indexOf('|');
  if (index === -1) {
    return ['', content.trim()];
  }
  return [content.slice(0, index).trim(), content.slice(index + 1).trim()];
};
