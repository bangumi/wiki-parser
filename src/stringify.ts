/* eslint-disable unicorn/no-array-reduce */
import { prefix, suffix } from './shared';
import type { Wiki, WikiArrayItem, WikiMap } from './types';

const stringifyArray = (arr: WikiArrayItem[] | undefined) => {
  if (!arr) {
    return '';
  }
  return arr.reduce((pre, item) => `${pre}\n[${item.k ? `${item.k}|` : ''}${item.v ?? ''}]`, '');
};

export function stringify(wiki: Wiki): string {
  const body = wiki.data.reduce((pre, item) => {
    if (item.array) {
      return `${pre}\n|${item.key} = {${stringifyArray(item.values)}\n}`;
    }
    return `${pre}\n|${item.key} = ${item.value ?? ''}`;
  }, '');

  const type = wiki.type ? ' ' + wiki.type : '';

  return `${prefix}${type}${body}\n${suffix}`;
}

export function stringifyMap(wiki: WikiMap): string {
  const body = [...wiki.data]
    .map(([_key, item]) => {
      if (item.array) {
        return `|${item.key} = {${stringifyArray(item.values)}\n}`;
      }

      return `|${item.key} = ${item.value ?? ''}`;
    })
    .join('\n');

  const type = wiki.type ? ' ' + wiki.type : '';

  return `${prefix}${type}\n${body}\n${suffix}`;
}
