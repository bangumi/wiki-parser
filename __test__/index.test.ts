import fs from 'node:fs';
import path from 'node:path';
import * as url from 'node:url';

import { describe, test, expect } from '@jest/globals';
import yaml from 'js-yaml';

import { parse, parseToMap, stringify, WikiArrayItem, WikiItem } from '../src';
import { stringifyMap } from '../src/stringify';

const __dirname = url.fileURLToPath(new URL('.', import.meta.url));

const testsDir = path.resolve(__dirname, 'wiki-syntax-spec/tests/');

const validTestDir = path.resolve(testsDir, 'valid');
const invalidTestDir = path.resolve(testsDir, 'invalid');

const validTestFiles = fs.readdirSync(validTestDir);
const inValidTestFiles = fs.readdirSync(invalidTestDir);

describe('Wiki syntax parser expected to be valid', () => {
  for (const file of validTestFiles) {
    const [prefix, suffix] = file.split('.');
    if (suffix !== 'wiki') {
      continue;
    }

    if (!prefix) {
      throw new Error('BUG: undefined file path prefix');
    }

    test(`${prefix} should be valid`, () => {
      const testFilePath = path.resolve(validTestDir, file);
      const expectedFilePath = path.resolve(validTestDir, `${prefix}.yaml`);

      const testContent = fs.readFileSync(testFilePath, 'utf8');
      const expectedContent = fs.readFileSync(expectedFilePath, 'utf8');

      const result = parse(testContent);
      const expected = yaml.load(expectedContent);

      expect(result).toEqual(expected);
    });
  }
});

describe('Wiki syntax parser expected to be inValid', () => {
  for (const file of inValidTestFiles) {
    const prefix = file.split('.')[0];
    if (!prefix) {
      throw new Error('BUG: undefined file path prefix');
    }

    test(`${prefix} should be invalid`, () => {
      const testFilePath = path.resolve(invalidTestDir, file);
      const testContent = fs.readFileSync(testFilePath, 'utf8');

      expect(() => parse(testContent)).toThrowError();
    });
  }
});

describe('Wiki stringify', () => {
  for (const file of validTestFiles) {
    const [prefix, suffix] = file.split('.');
    if (suffix !== 'wiki') {
      continue;
    }

    if (!prefix) {
      throw new Error('BUG: undefined file path prefix');
    }

    test(`${prefix} should stringify correct`, () => {
      const testFilePath = path.resolve(validTestDir, file);
      const expectedFilePath = path.resolve(validTestDir, `${prefix}.yaml`);

      const testContent = fs.readFileSync(testFilePath, 'utf8');
      const expectedContent = fs.readFileSync(expectedFilePath, 'utf8');
      const wiki = parse(testContent);
      const res = stringify(wiki);

      const result = parse(res);
      const expected = yaml.load(expectedContent);

      expect(result).toEqual(expected);
    });
  }
});

describe('parse map', () => {
  const item = new WikiItem('A', '', 'array');

  item.values = [new WikiArrayItem(undefined, 'b'), new WikiArrayItem(undefined, 'c')];

  const expected = { type: '', data: new Map([['A', item]]) };

  test('should merge keys', () => {
    const o = parseToMap(`{{Infobox
|A= b
| A=c
}}
`);

    expect(o).toEqual(expected);
  });

  test('should merge previous array keys', () => {
    const o = parseToMap(`{{Infobox
|A= {
[b]
}
| A=c
}}
`);

    expect(o).toEqual(expected);
  });

  test('should merge next array keys', () => {
    const parsed = parseToMap(`{{Infobox
|A= {
[1|b]
}
|c=e
| A=c
| A={
[2|b]
}
|q=
}}
`);

    expect(parsed).toMatchSnapshot();
    expect(stringifyMap(parsed)).toMatchSnapshot();
  });
});
