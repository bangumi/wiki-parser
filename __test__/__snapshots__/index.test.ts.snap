// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse map should merge next array keys 1`] = `
{
  "data": Map {
    "A" => WikiItem {
      "array": true,
      "key": "A",
      "value": undefined,
      "values": [
        WikiArrayItem {
          "k": "1",
          "v": "b",
        },
        WikiArrayItem {
          "k": undefined,
          "v": "c",
        },
        WikiArrayItem {
          "k": "2",
          "v": "b",
        },
      ],
    },
    "c" => WikiItem {
      "array": undefined,
      "key": "c",
      "value": "e",
      "values": undefined,
    },
    "q" => WikiItem {
      "array": undefined,
      "key": "q",
      "value": "",
      "values": undefined,
    },
  },
  "type": "",
}
`;

exports[`parse map should merge next array keys 2`] = `
"{{Infobox
|A = {
[1|b]
[c]
[2|b]
}
|c = e
|q = 
}}"
`;
